### PR TITLE
Fixes a big oversight with steal guns objective

### PR DESCRIPTION
**Pull request by Salex08**
## About The Pull Request

People with steal gun objectives like survivalist or thiefs would be able to greentext with items like toy guns, mining items or any other item which has its type as obj/item/gun

## Why It's Good For The Game

This PR adds a flag for unwanted guns which were not intended to be included in the steal guns objective and filters them out in a proc

## Changelog

:cl:@Salex08
fix: fixes a major oversight with steal gun objectives
/:cl:

### CODE REVIEW

**Comment by VladinXXV** [comment]
List of very much real guns that are excluded via this PR:
Russian Revolver (debatable)
Soul Russian Revolver (debatable)
Pipeguns
Regal Pipeguns
Miniature Energy Guns

**Comment by Salex08** [comment]
> List of very much real guns that are excluded via this PR: Russian Revolver (debatable) Soul Russian Revolver (debatable) Pipeguns Regal Pipeguns Miniature Energy Guns

thats the practice miniature energy gun not the real one and having pipeguns included which is craftable to steal is pretty nonsense in my opinion

**Comment by Salex08** [comment]
> List of very much real guns that are excluded via this PR: Russian Revolver (debatable) Soul Russian Revolver (debatable) Pipeguns Regal Pipeguns Miniature Energy Guns

**Comment by VladinXXV** [comment]
image
image

It is absolutely not the practice miniature energy gun. Just because you can craft something doesn't mean you can't steal it. Steal it from an assistant. Steal it from that one ghost role on icebox. Steal it from someone who happened to get one via the summon guns event.

**Comment by Salex08** [comment]
> image image
> 
> It is absolutely not the practice miniature energy gun. Just because you can craft something doesn't mean you can't steal it. Steal it from an assistant. Steal it from that one ghost role on icebox. Steal it from someone who happened to get one via the summon guns event.

yeah i just saw it now I meant /obj/item/gun/energy/e_gun/mini/practice_phaser not /obj/item/gun/energy/e_gun/mini I somehow forgot to add that, I accidently closed the PR when i had tried to tell you that

**Review by Mothblocks — CHANGES_REQUESTED** [review]
This should be a flag on the guns instead of a big ass list, especially since you seem to be including subtypes for no reason (typecaches should do that).

**Comment by Salex08** [comment]
> This should be a flag on the guns instead of a big ass list, especially since you seem to be including subtypes for no reason (typecaches should do that).

added a flag on the guns and made the filtering process different

File: code/game/gamemodes/objective.dm [hunk 1]
```diff
@@ -845,15 +845,22 @@ GLOBAL_LIST_EMPTY(possible_items_special)
...
+/datum/objective/steal_n_of_type/proc/check_if_valid_item(obj/item/I)
+	return TRUE
+
 /datum/objective/steal_n_of_type/summon_guns
 	name = "steal guns"
 	explanation_text = "Steal at least five guns!"
 	wanted_items = list(/obj/item/gun)
 	amount = 5
 
+/datum/objective/steal_n_of_type/summon_guns/check_if_valid_item(obj/item/gun/G)
```
Comment (Mothblocks) [inline code/game/gamemodes/objective.dm:861]: Keep the old variable name and type, we shouldn't change these very often.

Then check and cast inside the proc body itself.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/game/gamemodes/objective.dm [hunk 2]
```diff
@@ -845,15 +845,22 @@ GLOBAL_LIST_EMPTY(possible_items_special)
 		var/list/all_items = M.current.get_all_contents() //this should get things in cheesewheels, books, etc.
 		for(var/obj/I in all_items) //Check for wanted items
 			if(is_type_in_typecache(I, wanted_items))
-				stolen_count++
+				if(check_if_valid_item(I))
+					stolen_count++
 	return stolen_count >= amount
 
+/datum/objective/steal_n_of_type/proc/check_if_valid_item(obj/item/I)
```
[the remark this answers is not in the document]
Comment (Salex08) [reply code/game/gamemodes/objective.dm]: alright

File: code/game/gamemodes/objective.dm (hunk 1 above)
Comment (Salex08) [reply code/game/gamemodes/objective.dm]: done

**Pull request merged**